var/spack/repos/builtin/packages/gmsh/package.py

```diff
@@ -33,105 +36,127 @@ class Gmsh(CMakePackage):
     variant('petsc',       default=False, description='Build with PETSc')
     variant('slepc',       default=False, description='Build with SLEPc (only when PETSc is enabled)')
     variant('tetgen',      default=False, description='Build with Tetgen')
-    variant('metis',       default=False, description='Build with Metis')
+    variant('metis',       default=True,  description='Build with Metis')
     variant('privateapi',  default=False, description='Enable the private API')
+    variant('eigen',       default=False, description='Build with Eigen')
+    variant('cgns',        default=True,  description='Build with CGNS')
 
-    depends_on('blas')
-    depends_on('lapack')
+    # https://gmsh.info/doc/texinfo/gmsh.html#Compiling-the-source-code
+    # We make changes to the GMSH default, such as external blas.
+    depends_on('blas',    when='~eigen')
+    depends_on('lapack',  when='~eigen')
     depends_on('cmake@2.8:', type='build')
-    depends_on('gmp')
+    depends_on('gmp',  when='+gmp')
     depends_on('mpi',  when='+mpi')
     # Assumes OpenGL with GLU is already provided by the system:
-    depends_on('fltk', when='+fltk')
-    depends_on('hdf5', when='+hdf5')
-    depends_on('netgen', when='+netgen')
+    depends_on('fltk+gl', when='+fltk')
+    depends_on('cairo',   when='+cairo')
+    depends_on('hdf5',    when='+hdf5')
+    depends_on('mmg',     when='+mmg')
+    depends_on('netgen',  when='+netgen')
     depends_on('opencascade', when='+opencascade')
-    depends_on('oce',  when='+oce')
+    depends_on('oce',     when='+oce')
     depends_on('petsc+mpi', when='+petsc+mpi')
-    depends_on('petsc', when='+petsc~mpi')
-    depends_on('slepc', when='+slepc+petsc')
-    depends_on('tetgen', when='+tetgen')
-    depends_on('zlib',  when='+compression')
-    depends_on('metis', when='+metis')
+    depends_on('petsc',   when='+petsc~mpi')
+    depends_on('slepc',   when='+slepc+petsc')
+    depends_on('tetgen',  when='+tetgen')
+    depends_on('zlib',    when='+compression')
+    depends_on('metis',   when='+metis')
+    depends_on('eigen',   when='+eigen')
+    depends_on('cgns',    when='+cgns')
+    # Gmsh's high quality vector PostScript, PDF and SVG output is produced by GL2PS.
+    depends_on('gl2ps')
 
     conflicts('+slepc', when='~petsc')
     conflicts('+oce', when='+opencascade')
 
     def cmake_args(self):
         spec = self.spec
-        prefix = self.prefix
 
-        options = []
+        var_opt_list = [
+            ('+cairo', 'CAIRO'),
+            ('+cgns', 'CGNS'),
+            ('+fltk', 'FLTK'),
+            ('+gmp', 'GMP'),
+            ('+metis', 'METIS'),
+            ('+mpi', 'MPI'),
+            ('+openmp', 'OPENMP'),
+            ('+privateapi', 'PRIVATE_API'),
+        ]
+        # options setting copied from gribapi
+        options = ['-DENABLE_%s=%s' % (opt, 'ON' if var in spec else 'OFF')
+                   for var, opt in var_opt_list]
```

Comment:
`self.define_from_variant` helps here and protects against both typos and nonexistent variants:
```suggestion
        options = [
            self.define_from_variant('CAIRO'),
            self.define_from_variant('CGNS'),
            self.define_from_variant('FLTK'),
            self.define_from_variant('GMP') ,
            self.define_from_variant('METIS'),
            self.define_from_variant('MPI'),
            self.define_from_variant('OPENMP'),
            self.define_from_variant('PRIVATE_API', 'privateapi'),
        ]
```